src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractItemFileWatchDataSource.cs

```diff
@@ -0,0 +1,77 @@
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    /// <summary>
+    ///     Base class for producing items so that they can be watched, and when changed, trigger re-evaluation.
+    /// </summary>
+    internal abstract class AbstractItemFileWatchDataSource : ChainedProjectValueDataSourceBase<FileWatchData>, IFileWatchDataSource
+    {
+        private readonly ConfiguredProject _project;
+        private readonly IProjectSubscriptionService _projectSubscriptionService;
+
+        protected AbstractItemFileWatchDataSource(ConfiguredProject project, IProjectSubscriptionService projectSubscriptionService)
+            : base(project.Services, synchronousDisposal:true, registerDataSource: false)
+        {
+            _project = project;
+            _projectSubscriptionService = projectSubscriptionService;
+        }
+
+        public abstract string ItemRuleName
+        {
+            get;
+        }
+
+        public abstract FileWatchChangeKinds FileWatchChangeKinds
+        {
+            get;
+        }
+
+        protected override UnconfiguredProject ContainingProject
+        {
+            get { return _project.UnconfiguredProject; }
+        }
+
+        protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<FileWatchData>> targetBlock)
+        {
+            IProjectValueDataSource<IProjectSubscriptionUpdate> source = _projectSubscriptionService.ProjectRuleSource;
+
+            // Transform the changes from evaluation/design-time build -> restore data
+            DisposableValue<ISourceBlock<IProjectVersionedValue<FileWatchData>>> transformBlock = source.SourceBlock
+                                                                                                        .TransformWithNoDelta(update => update.Derive(u => CreateFileWatch(u.CurrentState)),
+                                                                                                            suppressVersionOnlyUpdates: true,
+                                                                                                            ruleNames: new[] { ItemRuleName });
+
+            // Set the link up so that we publish changes to target block
+            transformBlock.Value.LinkTo(targetBlock, DataflowOption.PropagateCompletion);
+
+            // Join the source blocks, so if they need to switch to UI thread to complete 
+            // and someone is blocked on us on the same thread, the call proceeds
+            JoinUpstreamDataSources(source);
+
+            return transformBlock;
+        }
+
+        private FileWatchData CreateFileWatch(IImmutableDictionary<string, IProjectRuleSnapshot> update)
+        {
+            IProjectRuleSnapshot snapshot = update.GetSnapshotOrEmpty(ItemRuleName);
+
+            var fullPaths = snapshot.Items.Select(item => item.Value)
+                                          .Select(metadata => metadata.GetValueOrDefault(Compile.FullPathProperty))
```

Comment:
Using `Compile` here is cheating. It would be better if this came from an abstract property, then new implementations are less likely to forget to add the `FullPath` property to their rule.

Reply:
hmm.. maybe this isn't really here any more? I can see its different in #5006